group/group.go

```diff
@@ -68,3 +73,73 @@ func (gc *GroupCommitment) Mul(scalar fr.Element) GroupCommitment {
 func (t GroupCommitment) Eq(cm GroupCommitment) bool {
 	return t.T_1.Equal(cm.T_1) && t.T_2.Equal(cm.T_2)
 }
+
+type MsmAccumulator struct {
+	g             Group
+	A_c           Element
+	baseScalarMap []msmCoeff
+}
+
+type msmCoeff struct {
+	basis  Element
+	scalar fr.Element
+}
+
+func NewMsmAccumulator(g Group) *MsmAccumulator {
+	return &MsmAccumulator{
+		g:             g,
+		A_c:           g.CreateElement(),
+		baseScalarMap: nil,
+	}
+}
+
+func (ma *MsmAccumulator) AccumulateCheck(
+	C Element,
+	scalar []fr.Element,
+	basis []Element,
+	rand *common.Rand) error {
+	if len(basis) != len(scalar) {
+		return fmt.Errorf("x and v must have the same length")
+	}
+
+	alpha, err := rand.GetFr()
+	if err != nil {
+		return fmt.Errorf("get random scalar: %s", err)
+	}
+
+	var tmp fr.Element
+outer:
+	for i := 0; i < len(basis); i++ {
+		tmp.Mul(&alpha, &scalar[i])
+
+		for j := range ma.baseScalarMap {
+			if ma.baseScalarMap[j].basis.Equal(basis[i]) {
+				var scalar fr.Element
+				scalar.Add(&ma.baseScalarMap[j].scalar, &tmp)
+				ma.baseScalarMap[j].scalar = scalar
+				continue outer
+			}
+		}
+		ma.baseScalarMap = append(ma.baseScalarMap, msmCoeff{basis: basis[i], scalar: tmp})
+	}
+	ma.A_c.AddAssign(C.ScalarMultiplication(C, alpha))
+
+	return nil
+}
+
+func (ma *MsmAccumulator) Verify() (bool, error) {
+	x := make([]fr.Element, 0, len(ma.baseScalarMap))
+	v := make([]Element, 0, len(ma.baseScalarMap))
+
+	for _, coeff := range ma.baseScalarMap {
+		v = append(v, coeff.basis)
+		x = append(x, coeff.scalar)
+	}
+
+	msmRes := ma.g.CreateElement()
+	if _, err := msmRes.MultiExp(v, x); err != nil {
+		return false, fmt.Errorf("computing msm: %s", err)
+	}
+
+	return msmRes.Equal(ma.A_c), nil
```

Comment:
All this is an adaptation of `msmaccumulator/msmaccumulator.go` but for the generic group abstraction.

It's mostly the same, but I couldn't use a `map` since using interfaces as keys doesn't really work in Go; so I had to use a slice.

Not a big deal; just mentioning the subtle difference.

Conceptually we still do the important idea: aggregate MSMs as random linear combinations if the basis of multiple ones overlap.